### PR TITLE
sekhonss Code Review 1

### PR DESCRIPTION
**Pull request by sukhbirsekhon**
#### Description
This android application allows users to share customizable playlists across different music platforms.

#### Suggestions:
- I noticed that the UI for the applications contains four different menu objects at the bottom. These are all separate buttons and there will be four different setOnClickListener for buttons which is not a recommended way of doing menu navigation. You should implement bottom navigation bar (from android) instead of setting up separate buttons for each page. It will take more time to load the page when calling each page using setOnClickListner because you are starting a new activity. Whereas, if you use the bottom navigation bar using fragments, it will save you memory allocation and better user experience. This is a suggestion from an efficiency point of view. 
- I also noticed that since you are using room database, you are initializing each DTO with an empty string so you might not get null pointer exceptions, but what about empty string problem? What if the user enters an empty song name? You would store that empty name inside the database. Basically, how are you handling empty strings? I spotted one on UI side where I can enter an empty string and the app would still make a playlist with no name. You should write if-else statements to make sure that the length of the string that the user enters should be greater than 0. Or you can also use isEmptyString() to validate the string. 
- My last suggestion would be to maintain Javadoc or commenting properly. I spotted a lot of places in the codebase where methods or functions lack comments. I added as many comments as much as I could. There were places where there were comments but lack explanations. For example: In the main activity, show alert dialog contains the code that is responsible for adding the playlist to the homepage and using lambdas functions to set the state of Dialog buttons (done or cancel), but there are no comments for that. My suggestion is to make sure in the future that you add comments as you finish writing each function because it is very important for other team members to also understand each other work. 

#### What I learned from this code review:
- I learned how to save data inside the room database by implementing Database and Entity annotations over the DTO data class objects while I was code reviewing the project. I also got to learn about Query and other database operation annotations (Insert and Delete). 

Code from codebase:
``` kotlin
@Query("SELECT * FROM playlists")
    fun getPlaylists(): LiveData<List<PlaylistDTO>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun savePlaylist(playlist: PlaylistDTO)

    @Delete
    fun deletePlaylist(playlist: PlaylistDTO)
```
- I learned what is 'suspend' and how to use it for asynchronous tasks. I also learned that there are three different types of dispatchers (UI, MAIN, DEFAULT) and all of them are used for different purposes. For example: if you are accessing the database, you should use Dispatchers.IO. 

Code from codebase: 

``` kotlin
private suspend fun savePlaylist(playlist: PlaylistDTO) {
        try {
            localPlaylistDAO.savePlaylist(playlist!!)
        } catch (e: Exception) {
            Log.e(TAG, e.message)
        }
    }
```
- I learned what is an 'abstract' keyword in front of the class and why to use it. You basically use abstract so that class can not be instantiated. I also learned that you can create abstract property and methods as well. 

Code from codebase:
``` kotlin
@Database(entities = [PlaylistDTO::class], version = 1)
abstract class AppDatabase: RoomDatabase() {
    abstract fun localPlaylistDAO(): ILocalPlaylistDAO
}
```
#### Comments:
- The application uses a room database to store the user selection of playlist which is the best choice for this type of application. Every time I opened the application, I would see my playlists in the application which really enhances the user experience. The backend of the program contains DTO and DAO for the playlist. It seems like the next step for this application is to implement the Songs DTO and DAO and add functionalities to make sure that the user can select songs. 
- The program was available in Github on time. 
- I found it a little harder to understand some parts of the program. For example, initially, I didn't know why AppDatabase class is using an abstract keyword. But for the most part, I think it was me who didn't know or understood some basic coding conventions. I can't say that program was well documented but it wasn't either bad. I would say the program is averagely documented.
- The program compiles.
- Most of my comments revolve around the basic Kotlin coding conventions, however, there were someplace where I actually dig in to make sure that logic is performing well. For example, an empty String problem was not taken care of so I added validation for that. I tried to understand the room database and make sure that everything is being used according to the best standards because I wanted to make this program more efficient. I looked up the best practices for everything in this program to add valuable commits. I also added a few commits for housekeeping stuff like removing clutter and unused or unnecessary code.

### CODE REVIEW

File: app/src/main/java/edu/uc/reedws/musiclink/IndividualPlaylistActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/IndividualPlaylistActivity.kt]: I removed the ' * ' from the imports because it is not a best practice to import all the imports under the library even if you don't need it. Let's say there is a java.awt.Event import which you didn't need but you still imported like this: java.awt.*. Now if you have your own Event import: reedws.edu.Event that you want to import, you won't be able to call Event object without writing the whole import location, which will lead to clutter in your codebase. 

You can also refer this community answer to understand more about import '*' issue here: https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt]: I removed the ' * ' from the imports because it is not a best practice to import all the imports under the library even if you don't need it. Let's say there is a java.awt.Event import which you didn't need but you still imported like this: java.awt.*. Now if you have your own Event import: reedws.edu.Event that you want to import, you won't be able to call Event object without writing the whole import location, which will lead to clutter in your codebase.

You can also refer this community answer to understand more about import '*' issue here: https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:52]: Housekeeping thing: added space to format a code little bit

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:46]: Housekeeping thing: added space to format a code little bit

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:40]: Housekeeping thing: added space to format a code little bit

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:59]: Housekeeping thing: removed space to format a code little bit

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:67]: Gave variable a meaningful name. editText doesn't say anything that it is, so I added switched the name from editText to searchedSong because that user search song using this text field.

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt]: I ran an application and I entered nothing inside the add playlist text field, but the application still added a playlist with no name instead of prompting the user that they have not entered anything. So I added a empty string validation check to make sure that the user enters something in the searched song field. If they do enter something, perform the regular operation, but if they don't, then create a toast that tells the user that they have not entered anything and do not create a playlist.

File: app/src/main/java/edu/uc/reedws/musiclink/SearchActivity.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/SearchActivity.kt]: I removed the ' * ' from the imports because it is not a best practice to import all the imports under the library even if you don't need it. Let's say there is a java.awt.Event import which you didn't need but you still imported like this: java.awt.*. Now if you have your own Event import: reedws.edu.Event that you want to import, you won't be able to call Event object without writing the whole import location, which will lead to clutter in your codebase.

You can also refer this community answer to understand more about import '*' issue here: https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

File: app/src/main/java/edu/uc/reedws/musiclink/dto/SongDTO.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/dto/SongDTO.kt]: It is always a good practice to have the ability to return a string of variables inside the data class to check the values inside the data class. I learned this from professor, but I have also looked it up in kotlin coding conventions. I am returning all the values from the data class, but you can tailored the string according to your need in the future.

Kotlin coding convention for data class: https://kotlinlang.org/docs/reference/data-classes.html

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt:23]: Added appropriate comments. Since this function is calling get playlist() which returns the LiveData list of playlist, therefore, I added a comments that it fetches the playlist which is added by the user.

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt:31]: Added an appropriate comment to indicate that this function creates and save new playlist inside room database

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt:33]: I changed the variable type from var to val because the playlist variable will not be changed further in the code. I learned from a professor that it is always a good practice to switch from var to val if the data is not set to be changed in the future.

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt:45]: Added an appropriate comment to indicate that this function saves or inserts the new playlist inside the room database because it calls savePlaylist() method.

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt]: Added an appropriate comment to indicate that this function retrieves all the playlists from the database because localPlaylistDAO calls the PlayListDTO which contains all the playlists.

File: app/src/main/java/edu/uc/reedws/musiclink/ui/main/ApplicationViewModel.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/ui/main/ApplicationViewModel.kt:25]: I moved the fetchPlaylists() down in the code because I learned from one of the professor videos (Sprint 1 code review) that private functions should be down below all the public functions in the class as it is one of the Kotlin coding conventions.

File: app/src/main/java/edu/uc/reedws/musiclink/ui/main/ApplicationViewModel.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/ui/main/ApplicationViewModel.kt:31]: I changed the visibility modifier of fetchPlaylists() to private because this function is only being used inside the ApplicationMainViewModel class and nowhere else. You can also refer to Kotlin visibility modifier coding convention here: https://kotlinlang.org/docs/reference/visibility-modifiers.html

File: app/src/main/java/edu/uc/reedws/musiclink/ui/main/MainFragment.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/ui/main/MainFragment.kt:16]: I removed the companion object from the top because I learned from one of the professor's lecture videos (Sprint 1 code review) that companion object should be down below in the class as it is one of the Kotlin coding conventions. You can also refer here: https://kotlinlang.org/docs/reference/code-style-migration-guide.html

File: app/src/main/java/edu/uc/reedws/musiclink/ui/main/MainFragment.kt
Comment (sukhbirsekhon) [inline app/src/main/java/edu/uc/reedws/musiclink/ui/main/MainFragment.kt:29]: I removed the companion object from the top because I learned from one of the professor's lecture videos (Sprint 1 code review) that companion object should be down below in the class as it is one of the Kotlin coding conventions. You can also refer here: https://kotlinlang.org/docs/reference/code-style-migration-guide.html

File: app/src/main/res/layout/add_playlist_dialog.xml
Comment (sukhbirsekhon) [inline app/src/main/res/layout/add_playlist_dialog.xml]: Changed the name of editText to searchedSong to make it more meaningful

File: app/src/test/java/edu/uc/reedws/musiclink/PlaylistLocalDatabaseTest.kt
Comment (sukhbirsekhon) [inline app/src/test/java/edu/uc/reedws/musiclink/PlaylistLocalDatabaseTest.kt:14]: Removed the unused imports to clear the clutter from the codebase. You might use it in the future, but I didn't see any use of these imports right now.

**Review by romigben — DISMISSED** [review]

File: app/src/main/java/edu/uc/reedws/musiclink/IndividualPlaylistActivity.kt
Comment (reediculous456) [reply app/src/main/java/edu/uc/reedws/musiclink/IndividualPlaylistActivity.kt]: @sukhbirsekhon I disagree with the * import thing. I agree that for 3rd party libraries * import is bad, but for our own activities it seems the Kotlin convention is to use * imports. If you run the code cleanup and reformat in android studio it will automatically convert activity imports to * imports

Comment (sukhbirsekhon) [reply app/src/main/java/edu/uc/reedws/musiclink/IndividualPlaylistActivity.kt]: @reediculous456 I think you have good point. I didn't know that when you clean up and reformat the code in Android studio, it converts imports to *. I never noticed it. I learned something new today.

Comment (sukhbirsekhon) [reply app/src/main/java/edu/uc/reedws/musiclink/IndividualPlaylistActivity.kt:8]: @reediculous456 I think you have a good point. I didn't know that when you clean up and reformat the code in Android studio, it converts imports to *. I never noticed it. I learned something new today.

**Review by reediculous456 — DISMISSED** [review]

**Review by romigben — DISMISSED** [review]

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (devagrawal09) [reply app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:67]: This is actually not the searched song, but the name of the playlist the user wants to create.
...and as I write this, I realize the point you are making haha

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (devagrawal09) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt]:
```suggestion
        val newPlaylistName = dialogLayout.findViewById<EditText>(R.id.newPlaylistName)
```

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (devagrawal09) [reply app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:67]: This is actually not the searched song, but the name of the playlist the user wants to create.
...and as I write this, I realize the point you are trying to make haha

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (devagrawal09) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt]:
```suggestion
            if (newPlaylistName.length() > 0) {
```

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (devagrawal09) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt]:
```suggestion
                    "You added " + newPlaylistName.text.toString(),
```

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (devagrawal09) [inline app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt]:
```suggestion
                val newPlaylistName = newPlaylistName.text
```

File: app/src/main/java/edu/uc/reedws/musiclink/dao/AppDatabase.kt
Comment (devagrawal09) [inline app/src/main/java/edu/uc/reedws/musiclink/dao/AppDatabase.kt:8]: I personally prefer only putting a space after the colon... not sure if this approach is somehow better. @reediculous456 @romigben what do you guys think, since we will be working on this code, and if I write something I will instinctively always put a space after and remove the space before.

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (devagrawal09) [inline app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt]:
```suggestion
     * Builds the database instance and retrieves the DAO instance from it
```

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (devagrawal09) [reply app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt]: This method actually only gets the DAO instance, it doesn't perform any database operations.
... again, I know, proving your point.

File: app/src/main/res/layout/add_playlist_dialog.xml
Comment (devagrawal09) [inline app/src/main/res/layout/add_playlist_dialog.xml]:
```suggestion
        android:id="@+id/newPlaylistName"
```

**Review by devagrawal09 — CHANGES_REQUESTED** [review]
@reediculous456 should we make and commit these changes before merging the PR? I can also make these changes later in Android Studio to make sure they don't break the build, or directly commit to this branch.

File: app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt
Comment (sukhbirsekhon) [reply app/src/main/java/edu/uc/reedws/musiclink/MainActivity.kt:67]: It's fine, I thought that the user can search for the song in the search bar and that is why I named the variable `searchedSong` that stores searched songs by a user.

**Comment by reediculous456** [comment]
> @reediculous456 should we make and commit these changes before merging the PR? I can also make these changes later in Android Studio to make sure they don't break the build, or directly commit to this branch.

@devagrawal09 we should get this branch to a point where it can be merged to master as the new code so directly commit to this branch but be sure to talk about the changes in the discussion

File: app/src/main/java/edu/uc/reedws/musiclink/dao/AppDatabase.kt
Comment (reediculous456) [reply app/src/main/java/edu/uc/reedws/musiclink/dao/AppDatabase.kt:8]: This is actually the Android Studio auto format that did this

Comment (discospiff) [reply app/src/main/java/edu/uc/reedws/musiclink/dao/AppDatabase.kt:8]: Control-Alt-L.

File: app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt
Comment (discospiff) [reply app/src/main/java/edu/uc/reedws/musiclink/service/PlaylistService.kt:23]: If adding KDoc, go all the way and describe the function as KDoc recommends: https://kotlinlang.org/docs/reference/kotlin-doc.html

File: app/src/test/java/edu/uc/reedws/musiclink/PlaylistLocalDatabaseTest.kt
Comment (discospiff) [reply app/src/test/java/edu/uc/reedws/musiclink/PlaylistLocalDatabaseTest.kt:14]: Control-Alt-O

**Comment by discospiff** [comment]
Several good suggestions here.  It looks like the group and reviewer are already having a conversation, so I'll leave it to the group to decide what to merge.

**Review by katievo-1 — APPROVED** [review]

**Pull request merged**
